magma/ast_utils.py

```diff
@@ -73,9 +73,44 @@ def wrapped(*args, **kwargs):
         return fn(enclosing_env, *args, **kwargs)
     return wrapped
 
+class NameCollector(ast.NodeVisitor):
+    def __init__(self):
+        self.names = set()
+
+    def visit_Name(self, node: ast.Name):
+        self.names.add(node.id)
+
+    def visit_FunctionDef(self, node: ast.FunctionDef):
+        self.names.add(node.name)
+
+    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef):
+        self.names.add(node.name)
+
+    def visit_ClassDef(self, node: ast.ClassDef):
+        self.names.add(node.name)
+
+def gen_free_name(tree: ast.AST, defn_env: dict, prefix: str = '__auto_name_'):
+    visitor = NameCollector()
+    visitor.visit(tree)
+    used_names = visitor.names | defn_env.keys()
+    f_str = prefix+'{}'
+    c = 0
+    name = f_str.format(c)
+    while name in used_names:
+        c += 1
+        name = f_str.format(c)
```

Comment:
Should we add a timeout (e.g. `if c >= 1000: raise Exception`). Unlikely to come up and would incur a performance cost, but you never know?

Reply:
As python ints don't overflow don't see the need to give up.

Reply:
I just realized we can assume that the space of names is finite, so we will eventually find a name.

Reply:
(because the length of the program must be finite)